0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+support efficient random access, and are more likely to be used incorrectly by accident:
+
+```dafny
+// Before:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  for i := 0 to |s| {
+    var left := s[i];
+    for j := 0 to |s| {
+      var right := s[i]; // Whoops!
+      result := result + (left, right);  
+    }
+  }
+}
+
+// After:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  foreach left | left in s {
+    foreach right | right in s {
+      result := result + (left, right);
+    }
+  }
+}
+
+// Or even better, as an expression:
+function AllPairs(s: seq<nat>): seq<(nat, nat)> {
+  seq left, right | left in s && right in s
+}
+```
+
+A more serious shortcoming is that there is currently no efficient way to iterate over the elements of a `set` at all.
+The best alternative is to use the assign-such-that operator and set subtraction, illustrated in the reference manual as follows
+([sec 10.5.2](https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#1052-sets)):
+
+```dafny
+// s is a set<int>
+var ss := s;
+while ss != {}
+  decreases |ss|
+{
+  var i: int :| i in ss;
+  ss := ss - {i};
+  print i, "\n";
+}
+```
+
+This is functionally correct, but because set subtraction requires creating a fresh copy,
+when compiled this loop requires quadratic time and memory (or at least garbage-collection work) in the size of the set.
+The only way to avoid this cost is to write additional external code in the target language to interface directly with the
+internal runtime implementation of `set` values.
+
+By comparison, the equivalent `foreach` loop is:
+
+```dafny
+foreach i | i in s {
+  print i, "\n";
+}
+```
+
+The runtime implementation of this loop can use the native iteration features of the target language.
+
+Going further, it is better to avoid writing any imperative loop at all where possible. Sticking to immutable data and
+expressions rather than statements allows logic to be used in specifications as well as implementation,
+and reduces the burden on the Dafny verifier. The proposed sequence comprehension expression allows more
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in a set is simple using a sequence comprehension expression: `seq i | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## foreach Loops
+
+A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
+but is substantially more general and expressive. A very simple example only looks slightly different than expected:
+
+```dafny
+foreach x | x in [1, 2, 3, 4, 5] {
+  print x, "\n";
+}
+```
+
+The feature supports much more sophisticated uses as well, however,
+including binding multiple variables at once and filtering:
+
+```dafny
+var myDoubleDeckerSeq: seq<seq<int>> := ...;
+foreach x, y | x in myDoubleDeckerSeq && y in x && y != 0 {
```

Comment:
That version would be rejected, actually, since the rules for `&&` say that the enumeration is nested left-to-right. `y != 0` is enumerable (as a sub-domain of `int`) but not finite, and then `y in x` for each value of `y` would not even be enumerable (i.e. all possible sequences that contain `y`).